src/main/java/com/aws/greengrass/authorization/AuthorizationModule.java

```diff
@@ -42,17 +47,55 @@ public void addPermission(String destination, Permission permission) throws Auth
                 || Utils.isEmpty(permission.getOperation())) {
             throw new AuthorizationException("Invalid arguments");
         }
-        // resource as null is ok, but it should not be empty
         String resource = permission.getResource();
-        if (resource != null && Utils.isEmpty(resource)) {
-            throw new AuthorizationException("Resource cannot be empty");
-        }
+        validateResource(resource);
         resourceAuthZCompleteMap.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                resource);
         rawResourceList.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                resource);
+    }
+
+    /**
+     * Only allow '?' if it's escaped. You can only escape special characters ('*', '$', '?').
+     * Any occurrence of '${' is only valid if it holds a single valid special character ('*', '$', '?') inside it
+     * and ends with '}'. (eg: "${*}" is valid, "${c}" is invalid, "${c" is invalid, ${*bc} is invalid)
+     * @param resource resource to be validated
+     */
+    private void validateResource(String resource) throws AuthorizationException {
+        if (resource == null) {
+            return;
+        }
+        // resource as null is ok, but it should not be empty
+        if (Utils.isEmpty(resource)) {
+            throw new AuthorizationException("Resource cannot be empty");
+        }
+        int length = resource.length();
+        for (int i = 0; i < length; i++) {
+            char currentChar = resource.charAt(i);
+            if (currentChar == escapeChar && i + 1 < length && resource.charAt(i + 1) == '{') {
+                char actualChar = getActualChar(resource.substring(i));
+                if (actualChar == nullChar) {
+                    throw new AuthorizationException("Resource contains invalid escape sequence. "
+                            + "Only ${*}, ${$}, or ${?} are allowed");
+                }
+                if (!isSpecialChar(actualChar)) {
+                    throw new AuthorizationException("Resource contains invalid escape "
+                            + "sequence: ${" + actualChar + "}. Only ${*}, ${$}, or ${?} are allowed");
```

Comment:
Can we change this to `"Resource contains an invalid escape sequence: ${<actualChar>}. You can use ${*}, ${$}, or ${?}."`